b9lab-content/5-my-own-chain/03-starport-13-game-wager.md

```diff
@@ -0,0 +1,271 @@
+storedGame := types.StoredGame{
+    ...
+    Wager: msg.Wager,
+}
+```
+And in the event:
+
+```go [https://github.com/cosmos/b9-checkers-academy-draft/blob/a8e8cdfe3f02697495f15d2348ed960635f32dc3/x/checkers/keeper/msg_server_create_game.go#L50]
+ctx.EventManager().EmitEvent(
+    sdk.NewEvent(sdk.EventTypeMessage,
+        ...
+        sdk.NewAttribute(types.StoredGameEventWager, strconv.FormatUint(msg.Wager, 10)),
+    )
+)
+```
+Less evidently, in order to avoid surprises down the road, you ought to also modify the creator function among the interface definition of `MsgCreateGame`, in `x/checkers/types/message_create_game.go`:
+
+```go [https://github.com/cosmos/b9-checkers-academy-draft/blob/a8e8cdfe3f02697495f15d2348ed960635f32dc3/x/checkers/types/message_create_game.go#L15]
+func NewMsgCreateGame(creator string, red string, black string, wager uint64) *MsgCreateGame {
+    return &MsgCreateGame{
+        ...
+        Wager: wager,
+    }
+}
+```
+
+## Preparing Your Keeper
+
+Saving a field named `Wager`, however well named, does not make players pay it or receive rewards. You need to add the handling actions. This handling has to ask the Bank to do the necessary. For this to be possible, your keeper needs to ask for a bank instance during setup. Remember that, because of the object-capability model throughout the Cosmos SDK, the only way to have access to a capability is to be given the reference to an instance that has this capability.
+
+What does your keeper need to do? It needs to keep in escrow the wager when the game is being played, and to pay back when the game is resolved. Good thing, the bank happens to have functions to transfer tokens from any account to your module, and vice-versa.
+
+Additionally, it is good practice to declare an interface that narrowly defines the functions you need for your purpose. The bank is capable of a lot but really, all you need here are 2 functions of bank. You _redeclare_ them in the file made for this purpose, `x/checkers/types/expected_keepers.go`:
+
+```go [https://github.com/cosmos/b9-checkers-academy-draft/blob/a8e8cdfe3f02697495f15d2348ed960635f32dc3/x/checkers/types/expected_keepers.go#L7-L10]
+type BankKeeper interface {
+    SendCoinsFromModuleToAccount(ctx sdk.Context, senderModule string, recipientAddr sdk.AccAddress, amt sdk.Coins) error
+    SendCoinsFromAccountToModule(ctx sdk.Context, senderAddr sdk.AccAddress, recipientModule string, amt sdk.Coins) error
+}
+```
+These 2 functions need to exactly match those declared [here](https://github.com/cosmos/cosmos-sdk/blob/8b78406/x/bank/keeper/keeper.go#L37-L39), copy / paste to the rescue. The magic of Go is that any object that has your 2 functions above will be a `BankKeeper`. Because the bank's keeper has these 2 functions, it will be one.
+
+Now with the requirements declared, it is time to make sure your keeper receives a reference. First, add a `BankKeeper` to your keeper in `x/checkers/keeper/keeper.go`:
```

Comment:
```suggestion
These two functions must exactly match the functions declared in the bank [keeper.go](https://github.com/cosmos/cosmos-sdk/blob/8b78406/x/bank/keeper/keeper.go#L37-L39) file. Copy and paste to the rescue. The magic of Go is that any object that has these two functions is a `BankKeeper`. Because the bank's keeper has these two functions, it will be a `BankKeeper`.

Now with the requirements declared, it is time to make sure your keeper receives a reference. First, add a `BankKeeper` to your keeper in `x/checkers/keeper/keeper.go`:
```